File: sdk/python/aragora_sdk/namespaces/marketplace.py
"""
Marketplace namespace for template discovery and deployment.

Provides API access to the template marketplace for discovering,
deploying, and managing workflow templates.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class MarketplaceAPI:
    """Synchronous marketplace API.

    Primary methods target FastAPI v2 marketplace routes. Legacy v1-only
    operations remain on v1 compatibility paths.
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    def list_templates(
        self,
        category: str | None = None,
        sort_by: Literal["downloads", "rating", "recent"] = "downloads",
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List available marketplace templates.

        Args:
            category: Filter by category
            sort_by: Sort order
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of templates with pagination
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if category:
            params["category"] = category
        # FastAPI v2 does not currently sort server-side, but accepts query
        # filtering/pagination. Keep sort_by for call-site compatibility.
        params["sort_by"] = sort_by

        return self._client.request("GET", "/api/v2/marketplace/templates", params=params)

    def search_templates(
        self,
        query: str,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        Search marketplace templates.

        Args:
            query: Search query
            limit: Maximum results
            offset: Pagination offset

        Returns:
            Matching templates
        """
        params: dict[str, Any] = {"q": query, "limit": limit, "offset": offset}
        return self._client.request("GET", "/api/v2/marketplace/templates", params=params)

    def search_templates_v1_compat(
        self,
        query: str,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Search templates using legacy compatibility route."""
        params: dict[str, Any] = {"query": query, "limit": limit, "offset": offset}
        return self._client.request("GET", "/api/marketplace/templates/search", params=params)

    def get_template(self, template_id: str) -> dict[str, Any]:
        """
        Get a template by ID.

        Args:
            template_id: Template identifier

        Returns:
            Template details
        """
        return self._client.request("GET", f"/api/v2/marketplace/templates/{template_id}")

    def get_template_reviews(
        self,
        template_id: str,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        Get ratings for a template.

        Args:
            template_id: Template identifier
            limit: Maximum ratings
            offset: Pagination offset

        Returns:
            Template ratings summary
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        return self._client.request(
            "GET", f"/api/v2/marketplace/templates/{template_id}/ratings", params=params
        )

    def deploy_template(
        self,
        template_id: str,
        name: str | None = None,
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Deploy a template to your workspace.

        Args:
            template_id: Template identifier
            name: Custom name for deployed workflow
            config: Template configuration overrides

        Returns:
            Deployment result with workflow_id
        """
        data: dict[str, Any] = {}
        if name:
            data["name"] = name
        if config:
            data["config"] = config

        return self._client.request(
            "POST", f"/api/v1/marketplace/templates/{template_id}/deploy", json=data
        )

    def get_deployment_status(self, deployment_id: str) -> dict[str, Any]:
        """
        Get deployment status.

        Args:
            deployment_id: Deployment identifier

        Returns:
            Deployment status
        """
        return self._client.request("GET", f"/api/v1/marketplace/deployments/{deployment_id}")

    def list_categories(self) -> dict[str, Any]:
        """
        List available template categories.

        Returns:
            List of categories
        """
        return self._client.request("GET", "/api/v2/marketplace/categories")

    def get_featured(self) -> dict[str, Any]:
        """
        Get featured templates.

        Returns:
            Featured templates list
        """
        return self._client.request("GET", "/api/v1/marketplace/featured")

    def submit_review(
        self,
        template_id: str,
        rating: int,
        comment: str | None = None,
    ) -> dict[str, Any]:
        """
        Submit a rating for a template.

        Args:
            template_id: Template identifier
            rating: Rating (1-5)
            comment: Review comment

        Returns:
            Rating submission confirmation
        """
        data: dict[str, Any] = {"score": rating}
        if comment:
            data["review"] = comment

        return self._client.request(
            "POST", f"/api/v2/marketplace/templates/{template_id}/ratings", json=data
        )

    def star_template(self, template_id: str) -> dict[str, Any]:
        """Star a marketplace template via FastAPI v2."""
        return self._client.request("POST", f"/api/v2/marketplace/templates/{template_id}/star")

    def export_template(self, template_id: str) -> dict[str, Any]:
        """Export a marketplace template via FastAPI v2."""
        return self._client.request("GET", f"/api/v2/marketplace/templates/{template_id}/export")

    def list_my_deployments(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List templates deployed to your workspace.

        Args:
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of deployments
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        return self._client.request("GET", "/api/v1/marketplace/my-deployments", params=params)

    def get_marketplace_status(self) -> dict[str, Any]:
        """
        Get marketplace status.

        GET /api/v2/marketplace/status

        Returns:
            Dict with marketplace status information
        """
        return self._client.request("GET", "/api/v2/marketplace/status")

    def get_marketplace_status_legacy(self) -> dict[str, Any]:
        """
        Get marketplace status via the legacy v1 compatibility route.

        GET /api/v1/marketplace/status

        Returns:
            Dict with marketplace status information
        """
        return self._client.request("GET", "/api/v1/marketplace/status")

    def get_circuit_breaker(self) -> dict[str, Any]:
        """
        Get marketplace circuit breaker status.

        GET /api/v1/marketplace/circuit-breaker

        Returns:
            Dict with circuit breaker status
        """
        return self._client.request("GET", "/api/v1/marketplace/circuit-breaker")


class AsyncMarketplaceAPI:
    """Asynchronous marketplace API.

    Primary methods target FastAPI v2 marketplace routes. Legacy v1-only
    operations remain on v1 compatibility paths.
    """

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    async def list_templates(
        self,
        category: str | None = None,
        sort_by: Literal["downloads", "rating", "recent"] = "downloads",
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List available marketplace templates."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if category:
            params["category"] = category
        # FastAPI v2 does not currently sort server-side, but accepts query
        # filtering/pagination. Keep sort_by for call-site compatibility.
        params["sort_by"] = sort_by

        return await self._client.request("GET", "/api/v2/marketplace/templates", params=params)

    async def search_templates(
        self,
        query: str,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Search marketplace templates."""
        params: dict[str, Any] = {"q": query, "limit": limit, "offset": offset}
        return await self._client.request("GET", "/api/v2/marketplace/templates", params=params)

    async def search_templates_v1_compat(
        self,
        query: str,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Search templates using legacy compatibility route."""
        params: dict[str, Any] = {"query": query, "limit": limit, "offset": offset}
        return await self._client.request("GET", "/api/marketplace/templates/search", params=params)

    async def get_template(self, template_id: str) -> dict[str, Any]:
        """Get a template by ID."""
        return await self._client.request("GET", f"/api/v2/marketplace/templates/{template_id}")

    async def get_template_reviews(
        self,
        template_id: str,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Get ratings for a template."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        return await self._client.request(
            "GET",
            f"/api/v2/marketplace/templates/{template_id}/ratings",
            params=params,
        )

    async def deploy_template(
        self,
        template_id: str,
        name: str | None = None,
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Deploy a template to your workspace."""
        data: dict[str, Any] = {}
        if name:
            data["name"] = name
        if config:
            data["config"] = config

        return await self._client.request(
            "POST", f"/api/v1/marketplace/templates/{template_id}/deploy", json=data
        )

    async def get_deployment_status(self, deployment_id: str) -> dict[str, Any]:
        """Get deployment status."""
        return await self._client.request("GET", f"/api/v1/marketplace/deployments/{deployment_id}")

    async def list_categories(self) -> dict[str, Any]:
        """List available template categories."""
        return await self._client.request("GET", "/api/v2/marketplace/categories")

    async def get_featured(self) -> dict[str, Any]:
        """Get featured templates."""
        return await self._client.request("GET", "/api/v1/marketplace/featured")

    async def submit_review(
        self,
        template_id: str,
        rating: int,
        comment: str | None = None,
    ) -> dict[str, Any]:
        """Submit a rating for a template."""
        data: dict[str, Any] = {"score": rating}
        if comment:
            data["review"] = comment

        return await self._client.request(
            "POST", f"/api/v2/marketplace/templates/{template_id}/ratings", json=data
        )

    async def star_template(self, template_id: str) -> dict[str, Any]:
        """Star a marketplace template via FastAPI v2."""
        return await self._client.request(
            "POST", f"/api/v2/marketplace/templates/{template_id}/star"
        )

    async def export_template(self, template_id: str) -> dict[str, Any]:
        """Export a marketplace template via FastAPI v2."""
        return await self._client.request(
            "GET", f"/api/v2/marketplace/templates/{template_id}/export"
        )

    async def list_my_deployments(
        self,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List templates deployed to your workspace."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        return await self._client.request(
            "GET", "/api/v1/marketplace/my-deployments", params=params
        )

    async def get_marketplace_status(self) -> dict[str, Any]:
        """Get marketplace status. GET /api/v2/marketplace/status"""
        return await self._client.request("GET", "/api/v2/marketplace/status")

    async def get_marketplace_status_legacy(self) -> dict[str, Any]:
        """Get marketplace status via the legacy v1 route. GET /api/v1/marketplace/status"""
        return await self._client.request("GET", "/api/v1/marketplace/status")

    async def get_circuit_breaker(self) -> dict[str, Any]:
        """Get marketplace circuit breaker status. GET /api/v1/marketplace/circuit-breaker"""
        return await self._client.request("GET", "/api/v1/marketplace/circuit-breaker")

File: sdk/typescript/src/namespaces/marketplace.ts
/**
 * Marketplace Namespace API
 *
 * Provides a namespaced interface for marketplace operations including
 * template publishing, discovery, rating, and deployment.
 */

import type { MarketplaceTemplate, PaginationParams, TemplateReview } from '../types';

/**
 * Template deployment status.
 */
export type DeploymentStatus = 'pending' | 'active' | 'paused' | 'archived' | 'failed';

/**
 * Template deployment information.
 */
export interface TemplateDeployment {
  id: string;
  template_id: string;
  tenant_id: string;
  name: string;
  status: DeploymentStatus;
  config: Record<string, unknown>;
  deployed_at: string;
  last_run?: string;
  run_count: number;
}

/**
 * Template rating summary.
 */
export interface TemplateRatings {
  ratings: Array<{ user_id: string; rating: number; created_at: string }>;
  average: number;
  count: number;
}

/**
 * Marketplace list parameters.
 */
export interface MarketplaceListParams {
  category?: string;
  search?: string;
  sort_by?: 'downloads' | 'rating' | 'newest';
  min_rating?: number;
  limit?: number;
  offset?: number;
}

export interface MarketplaceCatalogListParams {
  type?: string;
  tag?: string;
  category?: string;
  search?: string;
  limit?: number;
  offset?: number;
}

/**
 * Marketplace purchase record.
 */
export interface MarketplacePurchase {
  purchase_id: string;
  template_id: string;
  purchased_at: string;
  license_key?: string;
}

/**
 * Interface for the internal client methods used by MarketplaceAPI.
 */
interface MarketplaceClientInterface {
  request<T = unknown>(method: string, path: string, options?: Record<string, unknown>): Promise<T>;
  browseMarketplace(params?: MarketplaceListParams): Promise<{ templates: MarketplaceTemplate[] }>;
  getMarketplaceTemplate(templateId: string): Promise<MarketplaceTemplate>;
  publishTemplate(body: {
    template_id: string;
    name: string;
    description: string;
    category: string;
    tags?: string[];
    workflow_definition?: Record<string, unknown>;
    documentation?: string;
  }): Promise<{ marketplace_id: string }>;
  importTemplate(templateId: string, workspaceId?: string): Promise<{ imported_id: string }>;
  rateTemplate(templateId: string, rating: number): Promise<{ new_rating: number }>;
  reviewTemplate(templateId: string, body: {
    rating: number;
    title: string;
    content: string;
  }): Promise<{ review_id: string }>;
  getFeaturedTemplates(): Promise<{ templates: MarketplaceTemplate[] }>;
  getTrendingTemplates(): Promise<{ templates: MarketplaceTemplate[] }>;
  getMarketplaceCategories(): Promise<{ categories: string[] }>;
  getMarketplaceIndustries(): Promise<{ industries: string[] }>;
  getNewMarketplaceReleases(params?: { limit?: number }): Promise<{ templates: MarketplaceTemplate[] }>;
  getMarketplaceReviews(templateId: string, params?: PaginationParams): Promise<{ reviews: TemplateReview[] }>;
  purchaseTemplate(templateId: string, licenseType?: string): Promise<{ purchase_id: string; license_key?: string }>;
  downloadTemplate(templateId: string): Promise<{ content: Record<string, unknown>; version: string }>;
  getMyMarketplacePurchases(params?: PaginationParams): Promise<{ purchases: MarketplacePurchase[] }>;
  updateMarketplaceTemplate(templateId: string, body: {
    name?: string;
    description?: string;
    tags?: string[];
    documentation?: string;
    price?: number;
  }): Promise<MarketplaceTemplate>;
  unpublishTemplate(templateId: string): Promise<{ success: boolean }>;
  exportMarketplaceTemplate(templateId: string): Promise<Record<string, unknown>>;
}

/**
 * Marketplace API namespace.
 *
 * Provides methods for interacting with the Aragora template marketplace:
 * - Discovering and searching templates
 * - Publishing and managing templates
 * - Rating and reviewing templates
 * - Deploying templates to workspaces
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Browse featured templates
 * const { templates } = await client.marketplace.getFeatured();
 *
 * // Search for templates
 * const results = await client.marketplace.list({ category: 'analysis' });
 *
 * // Import a template to your workspace
 * const { imported_id } = await client.marketplace.import('template-123');
 *
 * // Rate a template
 * await client.marketplace.rate('template-123', 5);
 *
 * // Submit a review
 * await client.marketplace.review('template-123', {
 *   rating: 5,
 *   title: 'Great template!',
 *   content: 'Works perfectly for our use case.'
 * });
 *
 * // Publish your own template
 * const { marketplace_id } = await client.marketplace.publish({
 *   template_id: 'my-template',
 *   name: 'My Analysis Template',
 *   description: 'A template for analyzing data',
 *   category: 'analysis'
 * });
 * ```
 */
export class MarketplaceAPI {
  constructor(private client: MarketplaceClientInterface) {}

  // ===========================================================================
  // Discovery
  // ===========================================================================

  /**
   * List marketplace templates with optional filtering.
   *
   * @param params - Filter and pagination options
   * @returns List of templates matching the criteria
   */
  async list(params?: MarketplaceListParams): Promise<{ templates: MarketplaceTemplate[] }> {
    const response = await this.client.request<{
      templates: MarketplaceTemplate[];
      count?: number;
      limit?: number;
      offset?: number;
    }>('GET', '/api/v2/marketplace/templates', {
      params: {
        q: params?.search,
        category: params?.category,
        limit: params?.limit,
        offset: params?.offset,
      },
    });
    return { templates: response.templates };
  }

  /**
   * Get a specific template by ID.
   *
   * @param templateId - The template ID
   * @returns The template details
   */
  async get(templateId: string): Promise<MarketplaceTemplate> {
    return this.client.request<MarketplaceTemplate>('GET', `/api/v2/marketplace/templates/${encodeURIComponent(templateId)}`);
  }

  /**
   * Get featured templates curated by the Aragora team.
   *
   * @returns List of featured templates
   */
  async getFeatured(): Promise<{ templates: MarketplaceTemplate[] }> {
    return this.client.getFeaturedTemplates();
  }

  /**
   * Get trending templates based on recent activity.
   *
   * @returns List of trending templates
   */
  async getTrending(): Promise<{ templates: MarketplaceTemplate[] }> {
    return this.client.getTrendingTemplates();
  }

  /**
   * Get all marketplace categories.
   *
   * @returns List of category names
   */
  async getCategories(): Promise<{ categories: string[] }> {
    return this.client.request<{ categories: string[] }>('GET', '/api/v2/marketplace/categories');
  }

  /**
   * Get all marketplace industries.
   *
   * @returns List of industry names
   */
  async getIndustries(): Promise<{ industries: string[] }> {
    return this.client.getMarketplaceIndustries();
  }

  /**
   * Get newly released templates.
   *
   * @param limit - Maximum number of templates to return
   * @returns List of new templates
   */
  async getNewReleases(limit?: number): Promise<{ templates: MarketplaceTemplate[] }> {
    return this.client.getNewMarketplaceReleases({ limit });
  }

  /**
   * List marketplace catalog listings from the v1 pilot surface.
   */
  async listListings(params?: MarketplaceCatalogListParams): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/marketplace/listings', {
      params: {
        type: params?.type,
        tag: params?.tag,
        category: params?.category,
        search: params?.search,
        limit: params?.limit,
        offset: params?.offset,
      },
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * List featured marketplace catalog listings.
   */
  async listFeaturedListings(params?: { limit?: number }): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/marketplace/listings/featured', {
      params: { limit: params?.limit },
    }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get marketplace catalog listing stats.
   */
  async getListingStats(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/marketplace/listings/stats') as Promise<Record<string, unknown>>;
  }

  /**
   * Get marketplace catalog listing details.
   */
  async getListing(listingId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/v1/marketplace/listings/${encodeURIComponent(listingId)}`
    ) as Promise<Record<string, unknown>>;
  }

  /**
   * Search marketplace templates.
   *
   * @param params - Search parameters
   * @returns Search results
   */
  async search(params: {
    q?: string;
    category?: string;
  }): Promise<{ results: MarketplaceTemplate[]; total: number }> {
    const result = await this.list({ search: params.q, category: params.category });
    return { results: result.templates, total: result.templates.length };
  }

  // ===========================================================================
  // Reviews
  // ===========================================================================

  /**
   * Get reviews for a template.
   *
   * @param templateId - The template ID
   * @param params - Pagination parameters
   * @returns List of reviews
   */
  async getReviews(templateId: string, params?: PaginationParams): Promise<{ reviews: TemplateReview[] }> {
    const response = await this.client.request<{
      ratings: Array<{ user_id: string; score: number; review?: string; created_at: string }>;
    }>('GET', `/api/v2/marketplace/templates/${encodeURIComponent(templateId)}/ratings`, { params });
    return {
      reviews: response.ratings.map((rating, index) => ({
        review_id: `${templateId}-rating-${index + 1}`,
        template_id: templateId,
        user_id: rating.user_id,
        rating: rating.score,
        title: 'Rating',
        content: rating.review ?? '',
        created_at: rating.created_at,
      })),
    };
  }

  // ===========================================================================
  // Purchases & Downloads
  // ===========================================================================

  /**
   * Purchase a template from the marketplace.
   *
   * @param templateId - The template ID
   * @param licenseType - License type (default: 'standard')
   * @returns Purchase confirmation with ID and optional license key
   */
  async purchase(templateId: string, licenseType?: string): Promise<{ purchase_id: string; license_key?: string }> {
    return this.client.purchaseTemplate(templateId, licenseType);
  }

  /**
   * Download a purchased template.
   *
   * @param templateId - The template ID
   * @returns Template content and version
   */
  async download(templateId: string): Promise<{ content: Record<string, unknown>; version: string }> {
    return this.client.downloadTemplate(templateId);
  }

  /**
   * Get the current user's purchased templates.
   *
   * @param params - Pagination parameters
   * @returns List of purchases
   */
  async getMyPurchases(params?: PaginationParams): Promise<{ purchases: MarketplacePurchase[] }> {
    return this.client.getMyMarketplacePurchases(params);
  }

  // ===========================================================================
  // Publishing & Management
  // ===========================================================================

  /**
   * Publish a template to the marketplace.
   *
   * @param body - Template publication details
   * @returns The created marketplace ID
   */
  async publish(body: {
    template_id: string;
    name: string;
    description: string;
    category: string;
    tags?: string[];
    workflow_definition?: Record<string, unknown>;
    documentation?: string;
  }): Promise<{ marketplace_id: string }> {
    const response = await this.client.request<{ id: string; success: boolean }>('POST', '/api/v2/marketplace/templates', {
      body: {
        id: body.template_id,
        name: body.name,
        description: body.description,
        category: body.category,
        tags: body.tags,
        config: body.workflow_definition ?? {},
        documentation: body.documentation,
      },
    });
    return { marketplace_id: response.id };
  }

  /**
   * Import a template from the marketplace to your workspace.
   *
   * @param templateId - The template ID to import
   * @param workspaceId - Optional target workspace ID
   * @returns The imported template ID
   */
  async import(templateId: string, workspaceId?: string): Promise<{ imported_id: string }> {
    return this.client.importTemplate(templateId, workspaceId);
  }

  /**
   * Update an owned template in the marketplace.
   *
   * @param templateId - The template ID
   * @param body - Fields to update
   * @returns The updated template
   */
  async update(templateId: string, body: {
    name?: string;
    description?: string;
    tags?: string[];
    documentation?: string;
    price?: number;
  }): Promise<MarketplaceTemplate> {
    return this.client.updateMarketplaceTemplate(templateId, body);
  }

  /**
   * Unpublish (remove) an owned template from the marketplace.
   *
   * @param templateId - The template ID
   * @returns Success confirmation
   */
  async unpublish(templateId: string): Promise<{ success: boolean }> {
    return this.client.unpublishTemplate(templateId);
  }

  /**
   * Export a template as JSON.
   *
   * Useful for backing up templates or sharing outside the marketplace.
   *
   * @param templateId - The template ID to export
   * @returns The template definition as JSON
   *
   * @example
   * ```typescript
   * const templateJson = await client.marketplace.export('template-123');
   * // Save to file or process as needed
   * ```
   */
  async export(templateId: string): Promise<Record<string, unknown>> {
    return this.client.request<Record<string, unknown>>(
      'GET',
      `/api/v2/marketplace/templates/${encodeURIComponent(templateId)}/export`
    );
  }

  // ===========================================================================
  // Ratings & Reviews
  // ===========================================================================

  /**
   * Rate a template (1-5 stars).
   *
   * @param templateId - The template ID
   * @param rating - Rating value (1-5)
   * @returns The new average rating
   */
  async rate(templateId: string, rating: number): Promise<{ new_rating: number }> {
    if (rating < 1 || rating > 5) {
      throw new Error('Rating must be between 1 and 5');
    }
    const response = await this.client.request<{
      average_rating?: number;
      new_rating?: number;
    }>(
      'POST',
      `/api/v2/marketplace/templates/${encodeURIComponent(templateId)}/ratings`,
      { body: { score: rating } }
    );
    const newRating = response.average_rating ?? response.new_rating;
    if (typeof newRating !== 'number') {
      throw new Error('Marketplace rating response missing average_rating');
    }
    return { new_rating: newRating };
  }

  /**
   * Star a template in the marketplace.
   *
   * @param templateId - The template ID
   * @returns Updated star count
   */
  async star(templateId: string): Promise<{ success: boolean; stars: number }> {
    return this.client.request<{ success: boolean; stars: number }>(
      'POST',
      `/api/v2/marketplace/templates/${encodeURIComponent(templateId)}/star`
    );
  }

  /**
   * Submit a review for a template.
   *
   * @param templateId - The template ID
   * @param body - Review content
   * @returns The created review ID
   */
  async review(templateId: string, body: {
    rating: number;
    title: string;
    content: string;
  }): Promise<{ review_id: string }> {
    if (body.rating < 1 || body.rating > 5) {
      throw new Error('Rating must be between 1 and 5');
    }
    return this.client.reviewTemplate(templateId, body);
  }

  /**
   * Get marketplace status.
   */
  async getMarketplaceStatus(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v2/marketplace/status', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get marketplace status via the legacy v1 compatibility route.
   */
  async getMarketplaceStatusLegacy(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/marketplace/status', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get marketplace circuit breaker status.
   */
  async getCircuitBreaker(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/marketplace/circuit-breaker', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Search marketplace templates.
   *
   * @route GET /api/v2/marketplace/templates
   * @param params - Search parameters (query, category, tags, etc.)
   */
  async searchTemplates(params?: {
    q?: string;
    category?: string;
    tags?: string[];
    limit?: number;
    offset?: number;
  }): Promise<{ results: MarketplaceTemplate[]; total: number }> {
    const response = await this.client.request<{
      templates: MarketplaceTemplate[];
      count: number;
      limit: number;
      offset: number;
    }>('GET', '/api/v2/marketplace/templates', {
      params: {
        q: params?.q,
        category: params?.category,
        tags: params?.tags?.join(','),
        limit: params?.limit,
        offset: params?.offset,
      },
    });
    return { results: response.templates, total: response.count };
  }

  /**
   * Search marketplace templates via legacy compatibility route.
   *
   * @route GET /api/marketplace/templates/search
   */
  async searchTemplatesV1Compat(params?: {
    query?: string;
    category?: string;
    tags?: string[];
    limit?: number;
    offset?: number;
  }): Promise<{ results: MarketplaceTemplate[]; total: number }> {
    return this.client.request('GET', '/api/marketplace/templates/search', { params }) as Promise<{ results: MarketplaceTemplate[]; total: number }>;
  }

  /**
   * List the current user's template deployments.
   *
   * @route GET /api/v1/marketplace/my-deployments
   * @param params - Pagination and filter parameters
   */
  async listMyDeployments(params?: {
    status?: DeploymentStatus;
    limit?: number;
    offset?: number;
  }): Promise<{ deployments: TemplateDeployment[]; total: number }> {
    return this.client.request('GET', '/api/v1/marketplace/my-deployments', { params }) as Promise<{ deployments: TemplateDeployment[]; total: number }>;
  }
}
